Add RVL-CDIP dataset

dnaveenr opened the pull request:
Resolves #2762 

Dataset Request : Add RVL-CDIP dataset [#2762](https://github.com/huggingface/datasets/issues/2762)
This PR adds the RVL-CDIP dataset.

The dataset contains Google Drive link for download and wasn't getting downloaded automatically, so I have provided manual_download_instructions.
- I have added the dummy_data.zip as well.

Needed inputs on how I can run the real data and the dummy data tests for datasets with manual download ?

Inputs and suggestions for improvement are welcome. Thank you.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_4050). All of your documentation changes will be reflected on that endpoint.

mariosasko reviewed (commented):
Thanks! I've left a few comments.

You can use this URL to avoid manual download: `https://drive.google.com/uc?export=download&id=0Bz1dfcnrpXM-MUt4cHNzUEFXcmc`

Also, since the data files are TAR archives, we need to implement the streamable (can't use `os.path.join`) and the non-streamable versions of `_generate_examples` (see [this script](https://github.com/huggingface/datasets/blob/master/datasets/common_voice/common_voice.py) for instance). The labels-only data file URL doesn't work for me, so feel free to ask the authors whether they are OK with us hosting the file on the Hub/S3 (to speed up the streamable version). Let me know if you need help implementing this part.

dnaveenr commented:
Thanks a lot for input. I'll use the URL suggested and check.

>  we need to implement the streamable (can't use os.path.join) and the non-streamable versions of _generate_examples.

Sure. I will check the reference and try this out, will get back to you if I face any issues.

> The labels-only data file URL doesn't work for me, so feel free to ask the authors whether they are OK with us hosting the file on the Hub/S3 (to speed up the streamable version)

Just checked. The author (Adam Harley) has responded positively and allowed us to host the file. Do I share the file with you for hosting it on Hub/S3 ?

mariosasko commented:
> Just checked. The author (Adam Harley) has responded positively and allowed us to host the file. Do I share the file with you for hosting it on Hub/S3 ?

Yes, feel free to e-mail me the file. Then I'll create a repo under my namespace and push the file there. We run a GH action on a GH dataset after merging to create its repo on the Hub, so after this PR is merged, I'll push the file to the "official" namespace and update the download link.

dnaveenr commented:
Actually this URL works for direct download :
`https://drive.google.com/uc?export=download&confirm=pbef&id=0Bz1dfcnrpXM-MUt4cHNzUEFXcmc`
Ref : https://github.com/wkentaro/gdown/issues/146#issuecomment-1042382215

I'm working on the streaming download part as well, will update you regarding this.

mariosasko commented:
The rvl-cdip.tar.gz archive and txt files with the labels are on the Hub!

dnaveenr commented:
- Added 🤗 Hub download links.
- streamable and non-streamable versions of _generate_examples.
- Updated dummy data, both real and dummy dataset tests have passed.

mariosasko reviewed (commented):
Some minor nits, and we can merge then.

lhoestq commented on datasets/rvl_cdip/rvl_cdip.py:
```diff
@@ -0,0 +1,160 @@
+        archive_path = dl_manager.download(_URLS["rvl-cdip"])
+        labels_path = dl_manager.download(_METADATA_URLS)
+
+        # (Optional) In non-streaming mode, we can extract the archive locally to have actual local image files:
+        local_extracted_archive = dl_manager.extract(archive_path) if not dl_manager.is_streaming else None
```
I'm not a big fan of doing this, why do we need to extract it locally ? Can't we just use the extracted data from `dl_manager.iter_archive` ?

(for the record, we've added this feature for audio datasets because some audio formats can't easily be decoded from a fileobj and need actual files on disk - this is not the cause for images AFAIK)

dnaveenr replied:
Got it. This makes sense. I'll remove the extract part. :+1:

dnaveenr commented:
I've removed the extraction of the archive file locally as suggested. Let me know if any other changes are required. :)

mariosasko reviewed (approved):
I've pushed minor fixes/improvements to the card and the script. It all looks good now. Good job!

lhoestq reviewed (approved):
Thanks a lot @dnaveenr ! And thanks @mariosasko for updating the URLs

It looks all good to me :)

dnaveenr commented:
The check for **Update Hub repositories / update-hub-repositories** has failed.

> https://github.com/huggingface/datasets/runs/6116502392?check_suite_focus=true

lhoestq commented:
Hi ! Thanks for reporting ;) yes this CI job has been failing for a few days. I'm working on fixing it, and I'm manually running it on my side in the meantime

dnaveenr commented:
Great. :D Thank you @lhoestq
